controller/src/controller.rs

```diff
@@ -0,0 +1,149 @@
+}
+
+impl<T: BottlerocketNodeClient> BrupopController<T> {
+    pub fn new(node_client: T, brn_reader: Store<BottlerocketNode>) -> Self {
+        BrupopController {
+            node_client,
+            brn_reader,
+        }
+    }
+
+    /// Returns a list of all `BottlerocketNode` objects in the cluster.
+    fn all_nodes(&self) -> Vec<BottlerocketNode> {
+        self.brn_reader.state()
+    }
+
+    /// Returns the set of BottlerocketNode objects which is currently being acted upon.
+    ///
+    /// Nodes are being acted upon if they are not in the `WaitingForUpdate` state, or if their desired state does
+    /// not match their current state.
+    #[instrument(skip(self))]
+    fn active_node_set(&self) -> BTreeMap<String, BottlerocketNode> {
+        self.all_nodes()
+            .iter()
+            .filter(|brn| {
+                brn.status.as_ref().map_or(false, |brn_status| {
+                    brn_status.current_state != BottlerocketNodeState::WaitingForUpdate
+                        || !brn.has_reached_desired_state()
+                })
+            })
+            // kube-rs doesn't implement Ord or Hash on ObjectMeta, so we store these in a map indexed by name.
+            // (which are unique within a namespace). `name()` is guaranteed not to panic, as these nodes are all populated
+            // by our `reflector`.
+            .map(|brn| (brn.name(), brn.clone()))
+            .collect()
+    }
+
+    /// Determines next actions for a BottlerocketNode and attempts to execute them.
+    ///
+    /// This could include modifying the `spec` of a brn to indicate a new desired state, or handling timeouts.
+    #[instrument(skip(self), err)]
+    async fn progress_node(&self, node: BottlerocketNode) -> Result<()> {
+        if node.has_reached_desired_state() {
+            let desired_spec = determine_next_node_spec(&node);
+
+            event!(
+                Level::INFO,
+                ?desired_spec,
+                "BottlerocketNode has reached desired status. Modifying spec."
+            );
+
+            self.node_client
+                .update_node_spec(
+                    &node.selector().context(error::NodeSelectorCreation)?,
+                    &desired_spec,
+                )
+                .await
+                .context(error::UpdateNodeSpec)
+        } else {
+            // Otherwise, we need to ensure that the node is making progress in a timely fashion.
+
+            // TODO(seankell) Timeout handling will be added in a future PR.
+            Ok(())
+        }
+    }
+
+    /// This function searches all `BottlerocketNode`s for those which can be transitioned to a new state.
+    /// The state transition is then attempted. If successful, this node should be detected as part of the active
+    /// set during the next iteration of the controller's event loop.
+    #[instrument(skip(self))]
+    async fn search_for_node_to_update(&self) -> Option<BottlerocketNode> {
```

Comment:
This method secretly changed node's spec with a name seems only pick up some nodes.